src/hint_processor/builtin_hint_processor/ec_utils.rs

```diff
@@ -0,0 +1,562 @@
+pub fn recover_y_hint(
+    vm: &mut VirtualMachine,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+) -> Result<(), HintError> {
+    let p_x = get_integer_from_var_name("x", vm, ids_data, ap_tracking)?.into_owned();
+    let p_addr = get_relocatable_from_var_name("p", vm, ids_data, ap_tracking)?;
+    vm.insert_value(p_addr, &p_x)?;
+    let p_y = Felt::from(
+        recover_y(&p_x.to_biguint()).ok_or_else(|| HintError::RecoverYPointNotOnCurve(p_x))?,
+    );
+    vm.insert_value((p_addr + 1)?, p_y)?;
+    Ok(())
+}
+
+// Returns the Felt as a vec of bytes of len 32, pads left with zeros
+fn to_padded_bytes(n: &Felt) -> Vec<u8> {
+    let felt_to_bytes = n.to_bytes_be();
+    let mut bytes: Vec<u8> = vec![0; 32 - felt_to_bytes.len()];
+    bytes.extend(felt_to_bytes);
+    bytes
+}
+
+// Returns a random non-zero point on the elliptic curve
+//   y^2 = x^3 + alpha * x + beta (mod field_prime).
+// The point is created deterministically from the seed.
+fn random_ec_point_seeded(seed_bytes: Vec<u8>) -> Result<(Felt, Felt), HintError> {
+    // Hash initial seed
+    let mut hasher = Sha256::new();
+    hasher.update(seed_bytes);
+    let seed = hasher.finalize_reset().to_vec();
+    for i in 0..100 {
+        // Calculate x
+        let i_bytes = (i as u8).to_le_bytes();
+        let mut input = seed[1..].to_vec();
+        input.extend(i_bytes);
+        input.extend(vec![0; 10 - i_bytes.len()]);
+        hasher.update(input);
+        let x = BigUint::from_bytes_be(&hasher.finalize_reset());
+        // Calculate y
+        let y_coef = (-1).pow(seed[0] & 1);
+        let y = recover_y(&x);
+        if let Some(y) = y {
+            // Conversion from BigUint to BigInt doesnt fail
+            return Ok((Felt::from(x), Felt::from(y.to_bigint().unwrap() * y_coef)));
+        }
+    }
+    Err(HintError::RandomEcPointNotOnCurve)
+}
+const ALPHA: u32 = 1;
+lazy_static! {
+    static ref BETA: BigUint = BigUint::from_str_radix(
+        "3141592653589793238462643383279502884197169399375105820974944592307816406665",
+        10
+    )
+    .unwrap();
+}
+
+// Recovers the corresponding y coordinate on the elliptic curve
+//     y^2 = x^3 + alpha * x + beta (mod field_prime)
+//     of a given x coordinate.
+// Returns None if x is not the x coordinate of a point in the curve
+fn recover_y(x: &BigUint) -> Option<BigUint> {
+    let y_squared: BigUint = x.modpow(&BigUint::from(3_u32), &CAIRO_PRIME) + ALPHA * x + &*BETA;
+    if is_quad_residue(&y_squared) {
+        Some(sqrt(&Felt::from(y_squared)).to_biguint())
+    } else {
+        None
+    }
+}
+
+// Implementation adapted from sympy implementation
+// Conditions:
+// + prime is ommited as it will be CAIRO_PRIME
+// + a >= 0 < prime (other cases ommited)
+fn is_quad_residue(a: &BigUint) -> bool {
+    if a < &BigUint::from(2_u8) {
+        return true;
+    };
+    a.modpow(&(Felt::max_value().to_biguint() / 2_u32), &CAIRO_PRIME) == BigUint::one()
```

Comment:
This reduces heap allocation by 40%:
```suggestion
    a.is_zero() || a.is_one() ||
        a.modpow(&(Felt::max_value().to_biguint() / 2_u32), &CAIRO_PRIME).is_one()
```
Further reduction can be achieved (later on) by caching the result of `Felt::max_value().to_biguint() / 2_u32`, leaving it in only one allocation for the exponentiation.